client/protocol.py

```diff
@@ -0,0 +1,195 @@
+"""
+Protocol Client
+"""
+
+from __future__ import annotations
+import asyncio
+import logging
+import grpc
+
+from api.curp.message_pb2_grpc import ProtocolStub
+from api.curp.message_pb2 import FetchClusterRequest, FetchClusterResponse
+from api.curp.curp_command_pb2 import ProposeRequest, WaitSyncedRequest
+from api.xline.xline_command_pb2 import Command, CommandResponse, SyncResponse
+from client.error import ResDecodeError, CommandSyncError, WaitSyncError, ExecuteError
+from api.curp.curp_error_pb2 import (
+    CommandSyncError as _CommandSyncError,
+    WaitSyncError as _WaitSyncError,
+)
+from api.xline.xline_error_pb2 import ExecuteError as _ExecuteError
+
+
+class ProtocolClient:
+    """
+    Protocol client
+
+    Attributes:
+        leader_id: cluster `leader id`
+        connects: `all servers's `Connect`
+    """
+
+    leader_id: int
+    connects: dict[int, grpc.Channel]
+
+    def __init__(
+        self,
+        leader_id: int,
+        connects: dict[int, grpc.Channel],
+    ) -> None:
+        self.leader_id = leader_id
+        self.connects = connects
+
+    @classmethod
+    async def build_from_addrs(cls, addrs: list[str]) -> ProtocolClient:
+        """
+        Build client from addresses, this method will fetch all members from servers
+        """
+        cluster = await cls.fast_fetch_cluster(addrs)
+
+        connects = {}
+        for member in cluster.members:
+            channel = grpc.aio.insecure_channel(member.name)
+            connects[member.id] = channel
+
+        return cls(
+            cluster.leader_id,
+            connects,
+        )
+
+    @staticmethod
+    async def fast_fetch_cluster(addrs: list[str]) -> FetchClusterResponse:
+        """
+        Fetch cluster from server, return the first `FetchClusterResponse
+        """
+        futures = []
+        for addr in addrs:
+            channel = grpc.aio.insecure_channel(addr)
+            stub = ProtocolStub(channel)
+            futures.append(stub.FetchCluster(FetchClusterRequest()))
+        for t in asyncio.as_completed(futures):
+            return await t
+
+        msg = "fetch cluster error"
+        raise Exception(msg)
+
+    async def propose(self, cmd: Command, use_fast_path: bool = False) -> tuple[CommandResponse, SyncResponse | None]:
+        """
+        Propose the request to servers, if use_fast_path is false, it will wait for the synced index
+        """
+        if use_fast_path:
+            return await self.fast_path(cmd)
+        else:
+            return await self.slow_path(cmd)
```

Comment:
Should we use `try..catch` here to handle errors?

Reply:
I think it's not necessary for now. We can handle errors by catching them during function calls. After the mock server is completed, we will improve the error handling.